Sources/SwiftProtobuf/Message+StreamDecodingIterator.swift

```diff
@@ -0,0 +1,158 @@
+// Sources/SwiftProtobuf/Message+StreamDecodingIterator.swift - Iterator over binary delimited input streams
+//
+// Copyright (c) 2023 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Extensions to `Message` to provide an iterator for decoding binary delimited streams.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+fileprivate let defaultBufferLength = 32768
+
+extension Message {
+  
+  /**
+   Creates an iterator for a binary-delimited protobuf input stream. Errors can be caught by providing a
+   delegate. The returned iterator uses a read-ahead buffer, therefore no assumptions about the position of
+   the stream should be made after any messages have been read.
+   */
+  public static func streamDecodingIterator(inputStream: InputStream, bufferLength: Int? = nil, errorDelegate: StreamErrorDelegate? = nil) -> StreamDecodingIterator<Self> {
+    
+    let bufferLength = bufferLength ?? defaultBufferLength
+    return StreamDecodingIterator<Self>(inputStream: inputStream, bufferLength: bufferLength, errorDelegate: errorDelegate)
+  }
+}
+
+/**
+ Iterates over a binary-delimited protobuf input stream. This implementation uses a read-ahead buffer,
+ therefore after a message is read, the caller should assume any subsequent reads on the InputStream
+ will not begin at the start of a message.
+ */
+public struct StreamDecodingIterator<M: Message>: IteratorProtocol {
+  
+  fileprivate let buffer: ReadAheadBuffer
+  var errorDelegate: StreamErrorDelegate?
+  
+  init(inputStream: InputStream, bufferLength: Int, errorDelegate: StreamErrorDelegate?) {
+    self.buffer = ReadAheadBuffer(inputStream: inputStream, bufferLength: bufferLength)
+    self.errorDelegate = errorDelegate
+  }
+  
+  public func next() -> M? {
+    guard let messageLength = decodeVarint() else {
+      return nil
+    }
+    guard messageLength <= 0x7fffffff else {
+      errorDelegate?.onError(error: BinaryDecodingError.tooLarge)
+      return nil
+    }
+    if messageLength == 0 {
+      return M()
+    }
```

Comment:
@thomasvl Is this right for proto2 messages with required fields?  Seems that such messages could fail to decode with a zero-length input.

Reply:
Good catch, it is an edge case, but yea, if they are required, then this case should fail.  Looking back at BinaryDelimited.swift, we did add the `partial` arguments in there, so we could work those in here also to still get the initialization checks.

The catch is still the error flow - we don't want the iterator to fail and return `nil` because that indicates the end of the iteration, so how would that sorta failure get exposed?  Would it just skip on to the next message after reporting the error?